pkg/virtctl/vnc/vnc.go

```diff
@@ -69,75 +73,116 @@ func (o *VNC) Run(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	var opts k8smetav1.GetOptions
+	vmObj, err := virtCli.VM(namespace).Get(vm, opts)
+	if err != nil {
+		return fmt.Errorf("Can't access VM %s: %s", vm, err.Error())
+	}
+	if !vmObj.IsRunning() {
```

Comment:
OK, I see why you have to do it this way, with the current virtcli implementation. However, the server does that validation already. We should change the virtcli interface to allow you modelling  a proper flow.

As far as I can see, the issue is this `virtCli.VM(namespace).VNC(vm, pipeInReader, pipeOutWriter)` basically blocks until we are done. So you basically have no way to see if the underlying websocket connection can be established. Therefore you need to start remote-viewer in parallel already. Let's go one step back and have a look on how that is handle in k8s:

```golang
	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
	if err != nil {
		return "", err
	}

	err = exec.Stream(remotecommand.StreamOptions{
		Stdout: &stdout,
		Stderr: &stderr,
		Tty:    false,
	})
```

As you can see, there is a split between establishing the connection, and attaching streams to the connection. If we do that, you can simply first establish the connection, check for errors, and if there are none, you can start start a go-routine which attaches the streams which whe can also attach to remote-viewer.

A more useful interaction with streams for kubevirt may look like this:

```golang
errChan :- make(chan error)

vnc, err := virtCli.VM(namespace).VNC(vm)
if err != nil {
    return fmt.Errorf("could not establish connection: %v", err)
}

go func() {
  errChan  <- vnc.Stream(VNCStreamOptions{
		Stdout: &stdout,
		Stderr: &stderr,
	})
}()
```

Reply:
@rmohr great insight, that would improve the flow a lot. How would you like me to go forward? Should I file a different PR with the reworking of the streams or should I continue here?

Reply:
Please add it here. There is nothing better than having the improvement + the reason why the improvment is needed together :+1:

Reply:
@rmohr sure, working on it.

Reply:
hm, misread the code before a little bit. So the connection in the SPY case is created in stream.Stream(), however the principle stays the same. In there,  they first create a connection with the roundtripper, and then start the stream outside of the round-tripper.